### PR TITLE
Fix concurrent gauge creation

### PR DESCRIPTION
**Pull request by miklos-da**
Summary of changes:
* Introduced critical section when removing then registering a gauge such that in case multiple threads attempt to register the same gauge all of them will succeed.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by miklos-da** [comment]
> Very good catch Miklos.
> Would it not make sense to fix identical problem in CacheMetrics also at a fundamental level:
> https://github.com/digital-asset/daml/blob/d033f6560e69bce3d91c101adfc542cd79071194/ledger/metrics/src/main/scala/com/daml/metrics/CacheMetrics.scala#L26
> 
> 
> It could be in another PR.

Good point -- added that fix to this one.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 1]
```diff
@@ -8,7 +8,9 @@ import com.codahale.metrics._
 
 final class Metrics(val registry: MetricRegistry) {
 
-  private def gauge[T](name: MetricName, metricSupplier: MetricSupplier[Gauge[_]]): Gauge[T] = {
+  private[metrics] def gauge[T](
+      name: MetricName,
+      metricSupplier: MetricSupplier[Gauge[_]]): Gauge[T] = registry.synchronized {
     registry.remove(name)
     registry.gauge(name, metricSupplier).asInstanceOf[Gauge[T]]
   }
```
Comment (SamirTalwar-DA) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:16]:
```suggestion
  private[metrics] def gauge[T](name: MetricName, metricSupplier: MetricSupplier[Gauge[_]]): Gauge[T] =
    registry.synchronized {
      registry.remove(name)
      registry.gauge(name, metricSupplier).asInstanceOf[Gauge[T]]
    }
```

File: ledger/metrics/BUILD.bazel [hunk 2]
```diff
@@ -23,3 +24,14 @@ da_scala_library(
         "@maven//:io_dropwizard_metrics_metrics_jvm",
     ],
 )
+
+da_scala_test(
```
Comment (SamirTalwar-DA) [inline ledger/metrics/BUILD.bazel:28]: Can we use `da_scala_test_suite` here?

File: ledger/metrics/BUILD.bazel [hunk 3]
```diff
@@ -23,3 +24,14 @@ da_scala_library(
...
+    name = "metrics-tests",
+    size = "small",
+    srcs = glob(["src/test/**/*.scala"]),
```
Comment (SamirTalwar-DA) [inline ledger/metrics/BUILD.bazel:31]:
```suggestion
    srcs = glob(["src/test/scala/**/*.scala"]),
```

File: ledger/metrics/BUILD.bazel (hunk 2 above)
Comment (miklos-da) [reply ledger/metrics/BUILD.bazel]: Good point -- switched to that.

**Pull request merged**